### PR TITLE
Ccisse/state factory

### PR DESCRIPTION
**Pull request by ciss1995**

### CODE REVIEW

File: state/stateFactory.js [hunk 1]
```diff
@@ -0,0 +1,228 @@
...
+import { getEntityIdFromSirenEntity } from './sirenComponents/Common.js';
+import SirenParse from 'siren-parser';
+
+window.D2L = window.D2L || {};
+window.D2L.SirenSdk = window.D2L.SirenSdk || {};
+window.D2L.SirenSdk.fetchStatus = window.D2L.SirenSdk.fetchStatus || new Map();
+window.D2L.SirenSdk.d2lfetch = window.D2L.SirenSdk.d2lfetch || window.d2lfetch;
+window.D2L.SirenSdk.StateStore = window.D2L.SirenSdk.StateStore || new StateStore();
+
+export async function stateFactory(options) {
```
Comment (zinabat) [inline state/stateFactory.js:13]: You can use ES6 destructuring to do this more easily, e.g,
```js
export async function stateFactory({state, token, rawEntity, entityID}) {
    if (state) //etc
```
However, I'd like to see if we can make the usage for this more obvious/readable. There must be a way to have the function auto-detect when we are passing a half-fetched state vs an href and token vs a ready-made state and do what it needs to do without telling it explicitly.

**Review by zinabat — COMMENTED** [review]
First pass, I'll look at this more later

File: state/sirenComponents/ObserverMap.js [hunk 2]
```diff
@@ -0,0 +1,42 @@
+export class ObserverMap {
```
[the remark this answers is not in the document]
Comment (ciss1995) [reply state/sirenComponents/ObserverMap.js:1]: Oh yes good catch,  this is for another story.

File: state/stateFactory.js (hunk 1 above)
Comment (ciss1995) [reply state/stateFactory.js]: We could just have 2 vars one for token and one for the rest and check constructor.name and switch accordingly?

Comment (m6jones) [reply state/stateFactory.js]: Wait, why do we need state? This makes states.

File: framework/lit/HypermediaResult.js [hunk 3]
```diff
@@ -199,10 +199,10 @@ export class HypermediaResult extends TemplateResult {
 		const resources = { classes: [] };
 		const components = componentStoreFactory(pseudoTag);
 		if (!href || !token) return html`loading`;
-		const statePromise = stateFactory(href, token);
+		const statePromise = stateFactory(href.constructor.name, href, token);
 		const fetchedResults  = statePromise.then(async state => {
 			state.addObservables(resources, observable);
-			await fetch(state);
+			await stateFactory(state.constructor.name, state);
```
Comment (m6jones) [inline framework/lit/HypermediaResult.js:205]: This should still be a fetch call. We shouldn't need to call the stateFactory twice.

File: framework/lit/HypermediaStateMixin.js [hunk 4]
```diff
@@ -54,9 +54,9 @@ export const HypermediaStateMixin = superclass => class extends superclass {
 		if (this.__gettingState) return;
 		try {
 			this.__gettingState = true;
-			this._state = await stateFactory(this.href, this.token);
+			this._state = await stateFactory(this.href.constructor.name, this.href, this.token);
 			this._state.addObservables(this, this._observables);
-			await fetch(this._state);
+			await stateFactory(this._state.constructor.name, this._state);
```
Comment (m6jones) [inline framework/lit/HypermediaStateMixin.js:59]: Same this is still fetch.

File: state/sirenComponents/SirenLink.js [hunk 5]
```diff
@@ -63,11 +63,11 @@ export class SirenLink {
 		}
 
 		if (this._token) {
-			this._childState = await stateFactory(this.link.href, shouldAttachToken(this._token, this.link));
+			this._childState = await stateFactory(this.link.href.constructor.name, this.link.href, shouldAttachToken(this._token, this.link));
 			this._routes.forEach((route, component) => {
 				this._childState.addObservables(component, route);
 			});
-			fetch(this._childState);
+			stateFactory(this._childState.constructor.name, this._childState);
```
Comment (m6jones) [inline state/sirenComponents/SirenLink.js:70]: same this should remain fetch.

File: state/stateFactory.js [hunk 6]
```diff
@@ -0,0 +1,215 @@
...
+import { getEntityIdFromSirenEntity } from './sirenComponents/Common.js';
+import { HypermediaState } from './HypermediaState.js';
+import { StateStore } from './store.js';
+import SirenParse from 'siren-parser';
+
+window.D2L = window.D2L || {};
+window.D2L.SirenSdk = window.D2L.SirenSdk || {};
+window.D2L.SirenSdk.fetchStatus = window.D2L.SirenSdk.fetchStatus || new Map();
+window.D2L.SirenSdk.d2lfetch = window.D2L.SirenSdk.d2lfetch || window.d2lfetch;
+window.D2L.SirenSdk.StateStore = window.D2L.SirenSdk.StateStore || new StateStore();
```
Comment (m6jones) [inline state/stateFactory.js:12]: Lets remove SirenSdk from here.

File: state/stateFactory.js [hunk 7]
```diff
@@ -0,0 +1,215 @@
...
+import { getToken, shouldAttachToken } from './token.js';
+import { getEntityIdFromSirenEntity } from './sirenComponents/Common.js';
+import { HypermediaState } from './HypermediaState.js';
+import { StateStore } from './store.js';
+import SirenParse from 'siren-parser';
+
+window.D2L = window.D2L || {};
+window.D2L.SirenSdk = window.D2L.SirenSdk || {};
+window.D2L.SirenSdk.fetchStatus = window.D2L.SirenSdk.fetchStatus || new Map();
+window.D2L.SirenSdk.d2lfetch = window.D2L.SirenSdk.d2lfetch || window.d2lfetch;
```
Comment (m6jones) [inline state/stateFactory.js:11]: I don't think we need this.

File: state/stateFactory.js [hunk 8]
```diff
@@ -0,0 +1,215 @@
+import 'd2l-fetch/d2l-fetch.js';
+import { getToken, shouldAttachToken } from './token.js';
+import { getEntityIdFromSirenEntity } from './sirenComponents/Common.js';
+import { HypermediaState } from './HypermediaState.js';
+import { StateStore } from './store.js';
+import SirenParse from 'siren-parser';
+
+window.D2L = window.D2L || {};
+window.D2L.SirenSdk = window.D2L.SirenSdk || {};
+window.D2L.SirenSdk.fetchStatus = window.D2L.SirenSdk.fetchStatus || new Map();
```
Comment (m6jones) [inline state/stateFactory.js:10]: We don't need this either. fetchStatus will be held within fetchable classes.

File: state/stateFactory.js [hunk 9]
```diff
@@ -0,0 +1,215 @@
...
+window.D2L = window.D2L || {};
+window.D2L.SirenSdk = window.D2L.SirenSdk || {};
+window.D2L.SirenSdk.fetchStatus = window.D2L.SirenSdk.fetchStatus || new Map();
+window.D2L.SirenSdk.d2lfetch = window.D2L.SirenSdk.d2lfetch || window.d2lfetch;
+window.D2L.SirenSdk.StateStore = window.D2L.SirenSdk.StateStore || new StateStore();
+
+export async function stateFactory(type, objParam, token) {
+	switch (type) {
+		case 'HypermediaState':
+			return _fetch(objParam);
```
Comment (m6jones) [inline state/stateFactory.js:17]: stateFactory shouldn't fetch.

File: state/stateFactory.js [hunk 10]
```diff
@@ -0,0 +1,215 @@
...
+		case 'Link':
+			return _stateFactoryByRawSirenEntity(objParam, token);
+		case 'String':
+			return _stateFactory(objParam, token);
```
Comment (m6jones) [inline state/stateFactory.js:21]: HyepermediaState constructor can worry about if it was feed some data and how to get the entityId. We shouldn't need to tell it about what type.

File: state/stateFactory.js [hunk 11]
```diff
@@ -0,0 +1,215 @@
...
+		await _handleCachePriming(action.token, response);
+		const json = await response.json();
+		const entity = await SirenParse(json);
+		const state = window.D2L.SirenSdk.fetchStatus.get(entity.getLink('self').href, action.token);
+
+		state.onServerResponse(entity);
+	} catch (err) {
+		console.error(err);
+	}
+}
```
Comment (m6jones) [inline state/stateFactory.js:52]: This can be removed.

File: state/stateFactory.js [hunk 12]
```diff
@@ -0,0 +1,215 @@
...
+
+export async function refreshState(state, refetch = true) {
+	await state.refreshToken();
+	return _fetch(state, refetch);
+}
```
Comment (m6jones) [inline state/stateFactory.js:57]: Remove

**Review by m6jones — COMMENTED** [review]
It might be a better idea to go for the golden product but not connect it yet. I can connect it with the fetch.

So basically, the store needs to hold `HypermediaState` objects keyed by `token` (`token.toString`) and `href`

My screw up when I wrote the story :( Otherwise this was awesome!

File: state/stateFactory.js (hunk 10 above)
Comment (m6jones) [reply state/stateFactory.js:21]: My mind is arguing with this point. Maybe this is the place to parse out the entityId from an entity.

**Comment by ciss1995** [comment]
I believe this reflects what we want better https://github.com/BrightspaceHypermediaComponents/foundation-engine/pull/13

**Pull request closed without merging**